File: Basic/tutorial-applestamp/workflows/src/test/java/com/tutorial/RedeemApplesWithStampTest.java
package com.tutorial;

import com.google.common.collect.ImmutableList;
import com.tutorial.flows.CreateAndIssueAppleStamp;
import com.tutorial.flows.PackageApples;
import com.tutorial.flows.RedeemApples;
import com.tutorial.states.AppleStamp;
import com.tutorial.states.BasketOfApples;
import net.corda.core.contracts.UniqueIdentifier;
import net.corda.core.identity.CordaX500Name;
import net.corda.core.node.services.Vault;
import net.corda.core.node.services.vault.QueryCriteria;
import net.corda.core.transactions.SignedTransaction;
import net.corda.testing.node.*;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

public class RedeemApplesWithStampTest {
    private MockNetwork network;
    private StartedMockNode a;
    private StartedMockNode b;

    @Before
    public void setup() {
        network = new MockNetwork(new MockNetworkParameters().withCordappsForAllNodes(ImmutableList.of(
                TestCordapp.findCordapp("com.tutorial.contracts"),
                TestCordapp.findCordapp("com.tutorial.flows")))
                .withNotarySpecs(ImmutableList.of(new MockNetworkNotarySpec(CordaX500Name.parse("O=Notary,L=London,C=GB"))))
        );
        a = network.createPartyNode(null);
        b = network.createPartyNode(null);
        network.runNetwork();
    }

    @After
    public void tearDown() {
        network.stopNodes();
    }

    @Test
    public void buyerRedeemBasketOfApples() throws ExecutionException, InterruptedException {
        //Create Basket of Apples
        PackageApples.PackApplesInitiator createBasketOfApples = new PackageApples.PackApplesInitiator("Fuji4072", 10);
        Future<SignedTransaction> future = a.startFlow(createBasketOfApples);
        network.runNetwork();

        //Issue Apple Stamp
        CreateAndIssueAppleStamp.CreateAndIssueAppleStampInitiator issueAppleStamp =
                new CreateAndIssueAppleStamp.CreateAndIssueAppleStampInitiator(
                        "Fuji4072",this.b.getInfo().getLegalIdentities().get(0));
        Future<SignedTransaction> future1 = a.startFlow(issueAppleStamp);
        network.runNetwork();

        AppleStamp issuedStamp = (AppleStamp) future1.get().getTx().getOutputStates().get(0);
        UniqueIdentifier id = issuedStamp.getLinearId();

        //Redeem Basket of Apples with stamp
        RedeemApples.RedeemApplesInitiator redeemApples = new RedeemApples.RedeemApplesInitiator(b.getInfo().getLegalIdentities().get(0),id);
        Future<SignedTransaction> future2 = a.startFlow(redeemApples);
        network.runNetwork();

        //successful query means the state is stored at node b's vault. Flow went through.
        QueryCriteria outputCriteria = new QueryCriteria.VaultQueryCriteria().withStatus(Vault.StateStatus.UNCONSUMED);
        BasketOfApples state = b.getServices().getVaultService()
                .queryBy(BasketOfApples.class, outputCriteria).getStates().get(0).getState().getData();

        assert(state.getDescription().equals("Fuji4072"));
    }
}

File: Basic/tutorial-jarsigning/workflows/src/test/kotlin/com/tutorial/RedeemApplesWithStampTest.kt
package com.tutorial

import com.google.common.collect.ImmutableList
import com.tutorial.flows.CreateAndIssueAppleStampInitiator
import com.tutorial.flows.PackApplesInitiator
import com.tutorial.flows.RedeemApplesInitiator
import com.tutorial.states.AppleStamp
import com.tutorial.states.BasketOfApples
import net.corda.core.contracts.UniqueIdentifier
import net.corda.core.node.services.Vault.StateStatus
import net.corda.core.node.services.vault.QueryCriteria
import net.corda.core.transactions.SignedTransaction
import net.corda.testing.node.MockNetwork
import net.corda.testing.node.MockNetworkParameters
import net.corda.testing.node.StartedMockNode
import net.corda.testing.node.TestCordapp
import org.junit.After
import org.junit.Before
import org.junit.Test
import java.util.concurrent.ExecutionException
import java.util.concurrent.Future

class RedeemApplesWithStampTest {
    private var network: MockNetwork? = null
    private var a: StartedMockNode? = null
    private var b: StartedMockNode? = null
    @Before
    fun setup() {
        network = MockNetwork(MockNetworkParameters().withCordappsForAllNodes(ImmutableList.of(
                TestCordapp.findCordapp("com.tutorial.contracts"),
                TestCordapp.findCordapp("com.tutorial.flows"))))
        a = network!!.createPartyNode(null)
        b = network!!.createPartyNode(null)
        network!!.runNetwork()
    }

    @After
    fun tearDown() {
        network!!.stopNodes()
    }

    @Test
    @Throws(ExecutionException::class, InterruptedException::class)
    fun buyerRedeemBasketOfApples() {
        //Create Basket of Apples
        val createBasketOfApples = PackApplesInitiator("Fuji4072", 10)
        val future: Future<SignedTransaction> = a!!.startFlow(createBasketOfApples)
        network!!.runNetwork()

        //Issue Apple Stamp
        val issueAppleStamp = CreateAndIssueAppleStampInitiator(
                "Fuji4072", b!!.info.legalIdentities[0])
        val future1: Future<SignedTransaction> = a!!.startFlow(issueAppleStamp)
        network!!.runNetwork()
        val issuedStamp = future1.get().tx.outputStates[0] as AppleStamp
        val id = issuedStamp.linearId

        //Redeem Basket of Apples with stamp
        val redeemApples = RedeemApplesInitiator(b!!.info.legalIdentities[0], id)
        val future2: Future<SignedTransaction> = a!!.startFlow(redeemApples)
        network!!.runNetwork()

        //successful query means the state is stored at node b's vault. Flow went through.
        val outputCriteria: QueryCriteria = QueryCriteria.VaultQueryCriteria().withStatus(StateStatus.UNCONSUMED)
        val state = b!!.services.vaultService
                .queryBy(BasketOfApples::class.java, outputCriteria).states[0].state.data
        assert(state.description == "Fuji4072")
    }
}